### PR TITLE
Fixed GEMM example performance regression and batch GEMM fail issues.

### PR DESCRIPTION
**Pull request by jiyang1011**
hot fix : batch gemm & revert to previous config

pvc_gemm example performance:  216Tflops

### CODE REVIEW

**Comment by jiyang1011** [comment]
image
I also tried this config

**Review by taozha2 — APPROVED** [review]

**Review by aacostadiaz — APPROVED** [review]
Thanks for the PR. The issue with GEMM batch seems to be fixed. 

Regarding the performance regression, I am only able to get 156 TFLOPs with the configuration added in the PR. I tested different configurations which provided the correct output, but some of them are still failing and the performance is still significantly lower than what we had previously. Therefore, the performance regression will need to be addressed in a follow up PR.

**Comment by jiyang1011** [comment]
image
image
Please try this config.

File: include/cutlass/gemm/collective/xe_mma.hpp [hunk 1]
```diff
@@ -261,7 +267,9 @@ struct CollectiveMma<
         prefetch(mainloop.gmem_tiled_copy_b, iter_b(_,_,_,k_tile + DispatchPolicy::Stages));  
       }
     }
-    cute::gemm(tiled_mma, accum, tCrA, tCrB, src_accum);
+    for (int i = 0; i < SG_K / SubgroupSize; i++) {
```
Comment (mehdi-goli) [inline include/cutlass/gemm/collective/xe_mma.hpp:270]: Using the following configuration we are still getting the wrong result, even when using this PR.
```c++
using GmemTiledCopyA = XE_2D_U16x8x16_LD_N;
using GmemTiledCopyB = XE_2D_U16x16x16_LD_V;

using TileShape = Shape<_256, _256, _32>;

using TiledMma = TiledMMA<MMA_Atom<XE_8x16x16_F32BF16BF16F32_TT>,
       Layout<Shape<_4,_4,_2>>,
       Tile<_32,_64,_32>>;
```
by having a quick look into the code, it seems that the logic used in this for loop does not consider the   `Layout<Shape<_4,_4,_2>>` value when using the Atom `XE_8x16x16_F32BF16BF16F32_TT` to fulfil the computation of `Tile<_32,_64,_32>`. It seems that this for loop divide the K value with Subgroup-Size irrespective of K size in the thread group layout assuming it is always 1. So when the K size is the thread group layout is 2 (e.g. `Layout<Shape<_4,_4,_2>>`), it seems that we get a wrong result. Would you be able to investigate that?

Comment (jiyang1011) [reply include/cutlass/gemm/collective/xe_mma.hpp:270]: Thanks, your observation is right. Actually SG_K and BLK_K share the same parameter.
I am wondering whether  `Layout<Shape<_4,_4,_2>>,`  is  set for split-K algorithm ?

Comment (jiyang1011) [reply include/cutlass/gemm/collective/xe_mma.hpp:270]: When use `Layout<Shape<_4,_4,_2>>`, this means that only partial sum will be obtained ?

Comment (tdeng5) [reply include/cutlass/gemm/collective/xe_mma.hpp:270]: In this PR's, our goal is implementing the CUTE APIs which used in existing examples, and make sure there is no regression. How about merge this PR and make sure the CI/CD can work first. (BTW, the layout cannot work before this PR)

Yes, there are some corner cases we don't implement yet including the layout you mentioned, how about to file a JIRA/task to track these requirements?

**Review by tdeng5 — APPROVED** [review]

File: include/cutlass/gemm/collective/xe_mma.hpp [hunk 2]
```diff
@@ -200,8 +200,14 @@ struct CollectiveMma<
...
-    Tensor tCrA = thread_mma.partition_fragment_A(gA(_, _, 0));
-    Tensor tCrB = thread_mma.partition_fragment_B(gB(_, _, 0));
+    Tensor tCrA_partition = thread_mma.partition_fragment_A(gA(_, _, 0));
+    Tensor tCrA = make_tensor(static_cast<decltype(tCrA_partition) &&>(tCrA_partition).data(),
+                              tCrA_partition.shape());
+    Tensor tCrB_partition = thread_mma.partition_fragment_B(gB(_, _, 0));
+    Tensor tCrB = make_tensor(static_cast<decltype(tCrB_partition) &&>(tCrB_partition).data(),
+                              make_shape(size<0>(tCrB_partition.shape()),
+                                         size<2>(tCrB_partition.shape()),
+                                         size<1>(tCrB_partition.shape())));
```
Comment (mehdi-goli) [inline include/cutlass/gemm/collective/xe_mma.hpp:210]: I still don't get why we need to reshape the tensor here. Cannot find the logic behind it. We need to have a Jira ticket to track this and fix it later.

File: include/cutlass/gemm/collective/xe_mma.hpp (hunk 1 above)
Comment (mehdi-goli) [reply include/cutlass/gemm/collective/xe_mma.hpp:270]: Same here, generally this for loop should be handle inside the cut::gemm not here. Lets create a  ticket to address this separetly

**Review by mehdi-goli — APPROVED** [review]

**Pull request merged**